Overhaul client ping plumbing.

pchote opened the pull request:
This PR implements the next step towards server-side latency control by improving the ping measurements and routing the result through the main event queue where the future orderbuffer will be able to consume them and dial up/down the order latency as required. See the extended description of the main commit for more details.

The connection quality calculation is moved from the client to the server under the expectation that the new order buffer will be able to use the order stream to more accurately quantify the client's impact on the game quality, allowing us to finally add the connection indicators to the ingame player list.

To avoid scope creep and keep the momentum focused on the network code, i've resisted the temptation to rewrite the messy `PlayerPinger` code (this could be done in a separate PR) or to replace the flawed mean ping &rarr; quality calculation with something more robust (this is better left to the dynamic latency implementation).

This also moves the `NetTickScale` into `LobbyInfo.GlobalSettings` to remove another client-side dependency on old versus new netcode.

Depends on #19640.

pchote committed:
Rename connection packet handling.

pchote committed:
Move NetTickScale (now NetFrameInterval) control to the server.

teinarss commented on OpenRA.Game/Network/Session.cs:
```diff
@@ -122,6 +117,8 @@ public IEnumerable<Client> NonBotPlayers
 
 		public enum ClientState { NotReady, Invalid, Ready, Disconnected = 1000 }
 
+		public enum ConnectionQuality { Good, Moderate, Poor }
```
Not a fan of having these inside the class when they are also used outside this class, only makes the access a bit more cumbersome.

pchote replied:
I agree, but this is standard for everything in `Session` so it would be weird to *not* do that for this.

We should fix this more generally, but that would increase the scope of this PR by a factor of several.

jrb0001 commented on OpenRA.Game/Network/OrderIO.cs:
```diff
@@ -109,6 +109,27 @@ public static bool TryParseSync(byte[] packet, out (int Frame, int SyncHash, ulo
+		public static bool TryParsePing(int fromClient, byte[] packet, out byte[] ping)
+		{
+			// Valid Ping packets are only ever generated by the server
+			if (fromClient != 0 || packet.Length != 13 || packet[4] != (byte)OrderType.Ping)
```
If the client is supposed to treat this as opaque blob, then why not simply accept any length here?

pchote replied:
The idea is to minimize the chance of falsely matching random bytes as a valid packet so that we can fail correctly if something goes wonky in the network stream sending/parsing - this is already a problem with order packets (which we plan on fixing in a future PR), which caused me several headaches across this PR chain.

jrb0001 commented on OpenRA.Game/Network/UnitOrders.cs:
```diff
@@ -292,24 +292,20 @@ internal static void ProcessOrder(OrderManager orderManager, World world, int cl
+							{
+								var client = orderManager.LobbyInfo.Clients.FirstOrDefault(c => c.Index == int.Parse(strings[1]));
+								if (client != null)
+									client.ConnectionQuality = FieldLoader.GetValue<Session.ConnectionQuality>("clientPing", node.Value.Value);
```
Why `clientPing`?

jrb0001 reviewed (dismissed):
Looks good to me. But I think it would be more helpful to split between pure network latency (which the server can mitigate by delaying orders) and pure simulation time (which is a hard limit on how fast the server can trigger new ticks/frames without an uncontrolled increase in overall latency).

jrb0001 commented on OpenRA.Game/Network/Session.cs:
```diff
@@ -218,6 +218,9 @@ public class Global
 			public bool GameSavesEnabled;
 
+			// 120ms network frame interval for 40ms local tick
+			public int NetFrameInterval = 3;
```
Note that any value below 3 seems to cause slowdowns in singleplayer, because it uses hardcoded orderlatency 1 (Server.cs line 65) which (as far as I understand) causes stalls while waiting for the order to come back from the server (there is zero buffering).

pchote replied:
Yes, this will be addressed in future PRs by removing the hardcoded behaviour and allowing the dynamic latency system to handle it.

pchote commented:
Rebased and fixed the obsolete FieldLoader.GetValue field name.

> But I think it would be more helpful to split between

This doesn't fit with my current plans for finishing these protocol changes, so would delay things further without adding any immediate benefit. This could be done as a future PR after the limitations and requirements are clearer.

abcdefg30 commented on OpenRA.Game/Server/Server.cs:
```diff
@@ -1137,6 +1130,8 @@ public void SyncLobbyClients()
 				foreach (var t in serverTraits.WithInterface<INotifySyncLobbyInfo>())
 					t.LobbyInfoSynced(this);
+
+				pingUpdated.Restart();
```
Maybe we should put a comment here that this needs to change if we ever implement sending delta-changes (like mentioned in the TODO above).

abcdefg30 reviewed (commented):
`PlayerPinger` feels like a misnomer now, but that is OK for the time being I guess.

abcdefg30 reviewed (dismissed):
Lgtm once the comments are resolved.

pchote committed:
Overhaul client latency calculations.

The ping/pong orders are replaced with a dedicated
(and much smaller) Ping packet that is handled
directly in the client and server Connection wrappers.

This allows clients to respond when the orders are
processed, instead of queuing the pong order to be
sent in the next frame (which added an extra 120ms
of unwanted latency).

The ping frequency has been raised to 1Hz, and pings
are now routed through the server events queue in
preparation for the future dynamic latency system.

The raw ping numbers are no longer sent to clients,
the server instead evaluates a single ConnectionQuality
value that in the future may be based on more than
just the ping times.

pchote commented:
Done.

abcdefg30 commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/30da22eccb127dc751e67825dd40a14453c74fa8...fab3b860e3edd874d3cd962b6daa8b8705d739ed)
